pageserver/src/bin/pageserver.rs

```diff
@@ -200,23 +201,56 @@ fn initialize_config(
     })
 }
 
+struct WaitForPhaseResult<F: std::future::Future + Unpin> {
+    timeout_remaining: Duration,
+    skipped: Option<F>,
+}
+
+/// During startup, we apply a timeout to our waits for readiness, to avoid
+/// stalling the whole service if one Tenant experiences some problem.  Each
+/// phase may consume some of the timeout: this function returns the updated
+/// timeout for use in the next call.
+async fn wait_for_phase<F>(phase: &str, mut fut: F, timeout: Duration) -> WaitForPhaseResult<F>
+where
+    F: std::future::Future + Unpin,
+{
+    let initial_t = Instant::now();
+    let skipped = match tokio::time::timeout(timeout, &mut fut).await {
```

Comment:
Because we're retaining the future, we're not actually cancelling anything on the timeout AIUI.  It's true that it's not guaranteed that the caller nicely retains `skipped` and polls it later, but this function isn't file-scope and sufficient for the use case at hand.

(This should be equivalent to how timeouts were already being done for the init_logical_sizes phase)